modules/edu.gemini.seqexec.engine/src/test/scala/edu/gemini/seqexec/engine/packageSpec.scala

```diff
@@ -197,4 +200,41 @@ class packageSpec extends FlatSpec {
       ).unsafePerformSync._2.get._2.sequences(seqId3).status === SequenceState.Running
     )
   }
+
+  "engine" should "keep processing input messages regardless of how long Actions take" in {
+    val q = async.boundedQueue[Event](10)
+    val startedFlag = new Semaphore(0)
+    val finishFlag = new Semaphore(0)
+
+    val qs = Engine.State(Conditions.default,
+      None,
+      Map((seqId, Sequence.State.init(Sequence(
+        "First",
+        SequenceMetadata("GMOS", None),
+        List(
+          Step(
+            1,
+            None,
+            config,
+            Set(Resource.GMOS),
+            breakpoint = false,
+            List(
+              List(Task.apply{
+                startedFlag.release
+                finishFlag.acquire
+                Result.OK(Result.Configured("TCS"))
+              } )
+            )
+          )
+        )
+      ) ) ) )
+    )
+
+    val result = Nondeterminism[Task].both(
+        q.enqueueOne(start(seqId)) *> Task.apply(startedFlag.acquire) *>
+          q.enqueueOne(Event.getState{_ => Task.delay{finishFlag.release}}),
+        processE(q).drop(1).takeThrough(a => !isFinished(a._2.sequences(seqId).status)).run.eval(qs)
+      ).timed(5.seconds).unsafePerformSyncAttempt
```

Comment:
In general, I would avoid having timeouts in the tests. But in this case the wait only happens in the bad path, so I think it's OK.